design/annotation-for-power-cycling-and-deleting-failed-nodes.md

```diff
@@ -0,0 +1,188 @@
+- Provide the possibility of restoring cluster capacity automatically
+
+### Non-Goals
+
+- Creating a public API for rebooting a node
+- Identifying when automated recovery should be triggered
+- Taking actions other than power cycling to return the node to full health
+
+## Proposal
+
+This proposal calls for a new controller which watches for the presence of a
+specific Node annotation.  If present, the controller will locate the Machine
+and BareMetalHost host objects via their annotations, and (in serial) use the
+BareMetalHost API to power off the machine, delete the Node, and then power the
+machine back on again.
+
+In order to track the state of the recovery workflow, a [new
+CRD](https://github.com/kubevirt/machine-remediation/blob/master/pkg/apis/machineremediation/v1alpha1/machineremediation_types.go)
+is proposed.
+
+### User Stories
+
+#### Story 1
+
+As a HA component of Metal³ that has identified a failed node, I would like a
+declarative way to have the Node stopped; deleted; and restarted, so that I can
+recover exclusive workloads and restore cluster capacity.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The controller includes logic for power cycling a node.  
+
+As rebooting (software) and power cycling (hardware) a machine is a long
+running multi-step process, there are currently discussions around the creation
+of a public mechanism for requesting initiating these processes atomically.
+
+While it is likely that the work proposed here may make use of such
+functionality once it exists, it should not be considered a pre-requisite for
+the purposes of evaluating this proposal.
+
+### Risks and Mitigations
+
+RBAC rules will be needed to ensure that only specific roles can trigger
+machines to reboot. Without this, the system would be exposed to DoS style
+attacks.
+
+## Design Details
+
+See [PoC code](https://github.com/kubevirt/machine-remediation/)
+
+- A new [Machine Remediation CRD](https://github.com/kubevirt/machine-remediation/blob/master/pkg/apis/machineremediation/v1alpha1/machineremediation_types.go)
+- Two new controllers:
+  - [node reboot](https://github.com/kubevirt/machine-remediation/tree/master/pkg/controllers/nodereboot) which looks for the annoation and creates Machine Remediation CRs
+  - [machine remediation](https://github.com/kubevirt/machine-remediation/tree/master/pkg/controllers/machineremediation) which reboots the machine and deletes the Node object (which also erases the signalling annotation)
+- A new annotation (namespace and name is open for discussion)
+
+### Work Items
+
+- Make any requested modifications
+- Create a PR from  https://github.com/kubevirt/machine-remediation/ into https://github.com/metal3-io/baremetal-operator/
+
+### Dependencies
+
+This design is intended to integrate with OpenShift’s [Machine Healthcheck
+implementation](https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/machinehealthcheck/machinehealthcheck_controller.go#L407)
```

Comment:
So the machine health check controller notices a problem with a node and adds the annotation. Then the new node reboot controller sees the annotation and creates a machine remediation CR. Then the machine remediation controller uses that CR to power off the host, delete the node, and power the host back on.

Is that right?

Reply:
correct